### DISCUSSION TITLE
How to determine the correct n_boot parameter in predict interval?

### DISCUSSION THREAD

**Discussion by MarcosBelver** [JoaquinAmatRodrigo/skforecast#815] [Q&A]
I saw in the documentation that the n_boot parameter of the predict_interval function in the DirectForecaster is set to 500 by default.

I was wondering what is the mathematical/statistical explanation for this default selection. For example, if my set of residuals is ~150 (the training size is equal to that length) , does it make sense to oversampling the residuals obtaining a collection of 500 in order to get the corresponding quantile from there? Or this size of 500 is more oriented for longer training sets where the number of residuals is much bigger?

In that case, how to select the proper n_boot parameter for smaller residuals sample?

Thank you in advance.

**Comment by JoaquinAmatRodrigo** [comment]
Hi @MarcosBelver,
This is a good question.

Choosing 500 for n_boot strikes a practical balance between computational efficiency, accuracy and robustness. Increasing the number of bootstrap samples (larger n_boot) improves the approximation of the empirical distribution to the theoretical one. At around 500 samples, the distribution usually stabilises sufficiently to provide accurate quantile estimates, which are crucial for constructing robust prediction intervals. However, this is a rule of thumb based on only a few datasets and may not generalise to all cases.

As I understand it, having more iterations than residuals isn't inherently problematic because bootstrapping with replacement allows residuals to be reused. The main concern isn't "oversampling", but rather ensuring that the bootstrapped residuals accurately capture the variability of the original sample.

I hope this clarifies things. If you have any new insights, please feel free to share them - we are always keen to learn and improve the library!

**Comment by MarcosBelver** [comment]
Hi Joaquin!

Thank you for your explanations. I think makes more or less sense, however I do not see the advantages yet from the statistical perspective.

Assuming we are working with direct approach (with N models), shouldn't be enough to take directly the corresponding quantiles from the sample of residuals? Without any bootstrapping. What are the disadvantages you see to this method? It is like if you oversampling your residuals you will get approximately the same distribution

Look at the next picture

![image](https://github.com/user-attachments/assets/6de3867c-3269-4c16-b692-cd18904d0a18)

In my opinion, taking the same amount of samples than the size of the residuals the generated distribution more or less is equal to the expected one. Increasing this size does not contribute with too much information since what you are doing is to 'replicate' values. I could understand this from the perspective of a recursive model, but maybe not for the direct approach.

To sum it up:
- Why not take directly the corresponding quantiles from the sample of residuals (no bootstrap)? What are the disadvantages you can see here in your opinion?

- In the case of previous case is not applicable: wouldn't it be enough to establish n_boots ~ size of the residuals sample?

Again, thank you for your time and response. It is very interesting and helps to learn from other perspectives. 


On the other hand, I have seen you are planning to implement conformal prediction in the future. I am looking forward to this new feature in SkForecast. Recently I have been reading more about it and I think this could be the best solution to all the problems mentioned above. Of course, I know the implementation requires some time from your side and takes time.

Best regards,

**Comment by JoaquinAmatRodrigo** [comment]
Thank you, @MarcosBelver, for your detailed suggestion. 
I  agree with your point. It seems there’s no benefit to running the bootstrapping process; we can simply calculate the quantiles at each step directly. This approach will significantly speed up the process.

I will add this to the backlog for the next releases!
Regarding conformal intervals, we are getting familiar with this methodology and I hope we can add it soon.
